src/v/kafka/client/client.cc

```diff
@@ -295,18 +296,23 @@ client::create_topic(kafka::creatable_topic req) {
 
 ss::future<list_offsets_response>
 client::list_offsets(model::topic_partition tp) {
-    return gated_retry_with_mitigation([this, tp]() {
-        return _topic_cache.leader(tp)
-          .then(
-            [this](model::node_id node_id) { return _brokers.find(node_id); })
-          .then([tp](auto broker) mutable {
-              return broker->dispatch(kafka::list_offsets_request{
-                .data = {.topics{
-                  {{.name{std::move(tp.topic)},
-                    .partitions{
-                      {{.partition_index{tp.partition},
-                        .max_num_offsets = 1}}}}}}}});
-          });
+    using result = ss::future<list_offsets_response>;
+    return gated_retry_with_mitigation([this, _tp = std::move(tp)]() -> result {
+        auto me = this;
+        auto tp = _tp;
+        auto node_id = co_await me->_topic_cache.leader(tp);
+        auto broker = co_await me->_brokers.find(node_id);
+        auto res = co_await broker->dispatch(kafka::list_offsets_request{
+          .data = {.topics{
+            {{.name{tp.topic},
+              .partitions{
+                {{.partition_index{tp.partition}, .max_num_offsets = 1}}}}}}}});
+
+        auto ec = res.data.topics[0].partitions[0].error_code;
```

Comment:
Does this statement assume that there's at least one topic in the request? In theory a `list_offsets` request could contain no topics.

Reply:
It's my understanding that there is an equivalent `tp` in the response for every `tp` in the request.

I'm happy to add some more checks, though.

Reply:
Yeah. That's how I'm reading it too, which means that if the user provides an empty list of `tps` there's nothing at index 0.

Reply:
How would they request an empty list of `tps`?: `client::list_offsets(model::topic_partition tp) {`

Reply:
Fair point. I had looked at the `list_offsets_request` type which doesn't enforce a non empty list, but this invariant is established by the api it's fine.

Reply:
I agree that this shouldn't happen in practice, but I think it is worthwhile to check for empty topics and return an error: otherwise the remote node is able to crash us with a malformed response.

Reply:
Done